Add PeriodicFeatureExtractor

matthias-wende-frequenz opened the pull request:
A class for creating a profile from periodically occurring windows in a buffer of historical data that can be used for creating features for ml models.

closes #120

matthias-wende-frequenz committed:
Add PeriodicFeatureExtractor class

A class for creating a profile from periodically occurring windows in a
buffer of historical data that can be used for creating features for ml
models.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Add benchmark for the PeriodicFeatureExtractor

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Add methods to return timebounds from ringbuffer

This method is needed, such that it can be exposed by the `MovingWindow`
and eventually be used by the `TimeseriesAvg` to shift to be averaged
windows into the `MovingWindow`.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Add sampling_period property to MovingWindow

This is need in the `PeriodicFeatureExtractor` module to shift intervals
into the MovingWindow.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz commented:
The name should change in order to have a more consistent terminology (in line with MovingWindow). A maybe better name is `AvgWindow`, a tool that creates averaged windows.

@cwasicki, @idlir-shkurti-frequenz, @camille-bouvy-frequenz suggestions?

cwasicki commented:
Maybe something like `LagWindow`?

On another note, are you planning to support other metrics beyond `average` as well? Because we would also need e.g. `min`, `max`, `std`.

matthias-wende-frequenz committed:
Fix indexing in MovingWindow

When using a index slice where either the start or the stop key haven't
been defined these have been interpreted as None, which is the default
in the slice notation.

Instead using syntax where one is not set should be interpreted as an
interval starting with the first element or ending with the last
respectively.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Fix moving window unit tests

There have been inconsistencies w.r.t. the window alignment and how data
has been pushed.

Furthermore a unit test has been added that caused a bug in the unit
tests of the `PeriodicFeatureExtractor`.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

idlir-shkurti-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,137 @@
+    Initialize a PeriodicFeatureExtractor with a `MovingWindow` that contains no data.
+
+    Args:
+        period: List[float]
```
Please define what the period is here.

matthias-wende-frequenz replied:
done :)

matthias-wende-frequenz committed:
Add tests for the PeriodicFeatureExtractor

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz committed:
Make some methods in ringbuffer public

We make the `wrap` and `normalize_timestamp` methods public such that
they can be used by users of the ringbuffer.

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,471 @@
+
+        return resized_array.reshape(num_windows, self._distance)
+
+    def _calculate_avg_window(
```
This method is no longer used except in the benchmark and should be moved to the benchmark.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,471 @@
+            reshaped[:, :window_size], axis=0, weights=weights
+        )
+
+    def _calculate_avg_window_py(
```
This method is no longer used except in the benchmark and should be moved to the benchmark.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,471 @@
+
+        if start_pos >= end_pos:
+            window_start = self._moving_window[start_pos:]
+            window_end = self._moving_window[:end_pos]
```
Don't use the moving windows [] operator for data access, as the semantics are planned to change soon. As an alternative we can use the ringbuffer directly as we have the instance as `self._buffer` available.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -204,6 +217,24 @@ async def sink_buffer(sample: Sample) -> None:
+        Returns:
+            The timestamp of the newest sample of the ring buffer.
+        """
+        return self._buffer.time_bound_newest()
```
These are good candidates for properties too:

```suggestion
    @property
    def time_bound_oldest(self) -> datetime:
        """
        Return the time bounds of the ring buffer.

        Returns:
            The timestamp of the oldest sample of the ring buffer.
        """
        return self._buffer.time_bound_oldest()

    @property
    def time_bound_newest(self) -> datetime:
        """
        Return the time bounds of the ring buffer.

        Returns:
            The timestamp of the newest sample of the ring buffer.
        """
        return self._buffer.time_bound_newest()
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_ringbuffer/buffer.py:
```diff
@@ -152,6 +152,24 @@ def update(self, sample: Sample) -> None:
+        Returns:
+            The timestamp of the newest sample of the ring buffer.
+        """
+        return self._datetime_newest
```
These are good candidates for properties:

```suggestion
    @property
    def time_bound_oldest(self) -> datetime:
        """
        Return the time bounds of the ring buffer.

        Returns:
            The timestamp of the oldest sample of the ring buffer.
        """
        return self._datetime_oldest

    @property
    def time_bound_newest(self) -> datetime:
        """
        Return the time bounds of the ring buffer.

        Returns:
            The timestamp of the newest sample of the ring buffer.
        """
        return self._datetime_newest
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -163,6 +166,16 @@ def __init__(  # pylint: disable=too-many-arguments
+    @property
+    def sampling_period(self) -> timedelta:
+        """
+        Return the sampling period of the MovingWindow.
+
+        Returns:
+            The sampling period of the MovingWindow.
```
Nitpicking :grimacing: 

```suggestion
        Return the sampling period of the `MovingWindow`.

        Returns:
            The sampling period of the `MovingWindow`.
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_moving_window.py:
```diff
@@ -261,7 +292,21 @@ def __getitem__(self, key: SupportsIndex | datetime | slice) -> float | ArrayLik
+                    key = slice(0, key.stop)
+                if key.stop is None:
+                    _logger.debug("Setting stop to last index %s", self.__len__())
+                    key = slice(key.start, self.__len__())
```
Not sure if you can apply it here, but there is a fairly undocumented method of `slice` that will convert the slice `None` values to real values if you give it a `len` for the container:

```console
$ pydoc slice.indices |cat -p
Help on method_descriptor in slice:

slice.indices = indices(...)
    S.indices(len) -> (start, stop, stride)
    
    Assuming a sequence of length len, calculate the start and stop
    indices, and the stride length of the extended slice described by
    S. Out of bounds indices are clipped in a manner consistent with the
    handling of normal slices.

```

Some usage examples:

```pycon
>>> slice(5).indices(10)
(0, 5, 1)
>>> slice(20).indices(10)
(0, 10, 1)
>>> slice(1, 10).indices(10)
(1, 10, 1)
>>> slice(1, 15).indices(10)
(1, 10, 1)
>>> slice(1, -1).indices(10)
(1, 9, 1)
>>> slice(0, -1).indices(10)
(0, 9, 1)
>>> slice(-5, -1).indices(10)
(5, 9, 1)
```

It won't probably work for the wrapping around, but it might help to reduce the boilerplate here.

matthias-wende-frequenz replied:
Unfortunately that works only for integers, but at least we can reduce the code by two lines.

leandro-lucarella-frequenz replied:
Amazing! :stuck_out_tongue:

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
If this needs to expose only one class, I would make it a private module (`_periodic_feature_extractor.py`) and expose the class in `frequenz.sdk.timeseries` (`src/frequenz/sdk/timeseries/__init__.py`).

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+import numpy as np
+from numpy.typing import NDArray
+
+from frequenz.sdk.timeseries._moving_window import MovingWindow
```
```suggestion
from ._moving_window import MovingWindow
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+
+from frequenz.sdk.timeseries._moving_window import MovingWindow
+
+log = logging.getLogger(__name__)
```
We had some PR recently changing all logger instances to `_logger` :)

```suggestion
_logger = logging.getLogger(__name__)
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+This class calculates a statistical profile S over all $S_i$,
+i.e. the value of S at position i is calculated by performing
+a certian calculation, e.g. an average, over all values of
+$S_i$ at position i.
```
Some format improvements. And I replaced the `S` in `profile S over` and `value of S at position` to `$S$`, not sure if that's correct.

```suggestion
Example:

    Consider a timeseries $T$ of historical data and
    sub-series $S_i$ of $T$ all having the same size $l$.
    and the same fixed distance $p$ called period,
    where period of two sub-windows is defined as the distance
    of two points at the same position within the sub-windows.

    This class calculates a statistical profile $S$ over all $S_i$,
    i.e. the value of $S$ at position i is calculated by performing
    a certain calculation, e.g. an average, over all values of
    $S_i$ at position i.
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+    start and end time modulo a fixed period.
+
+    Note that the oldest window or the window that is currently overwritten in
+    the MovingWindow is not considered in the profile.
```
```suggestion
    the `MovingWindow` is not considered in the profile.
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+    Note that the oldest window or the window that is currently overwritten in
+    the MovingWindow is not considered in the profile.
+
+    Note further that when constructing a PeriodicFeatureExtractor object
```
```suggestion
    Note further that when constructing a `PeriodicFeatureExtractor` object
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+            distance=timedelta(days=7),
+        )
+
+        now = datetime.now(tzinfo=timezone.utc)
```
This is just a FYI because I've seen many people use `datetime.now(tzinfo=...)`. You don't actually need to use a keyword argument here, you can just write `datetime.now(timezone.utc)`. Since the value already says `timezone` in it, I think it is pretty clear that is the timezone we are passing (and the type checking will make sure the type is right also).

I'm fine with leaving the keyword argument, again, just a FYI in case it wasn't clear.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        )
+
+        # create a daily average for Thursday March 23 2023
+        th_avg_24h = feature_extractor(datetime(2023, 03, 23), datetime(2023, 03, 24))
```
```suggestion
        th_avg_24h = feature_extractor.avg(datetime(2023, 03, 23), datetime(2023, 03, 24))
```

?

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+
+    The profile is created out of all windows that are fully
+    contained in the underlying buffer with the same
+    start and end time modulo a fixed period.
```
where do the start and end time come from?  is it from the MovingWindow?  Because the text doesn't say MovingWindow, but "underlying buffer".

matthias-wende-frequenz replied:
yes underlaying buffer is the moving window. this class can (in theory) work with a numpy array too.

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+
+    Note further that when constructing a PeriodicFeatureExtractor object
+    the `MovingWindow` size has to be a integer multiple of
+    the period.
```
There are a lot of inconsistently in paragraph formatting.  Sometimes they are very narrow,  sometimes too wide.  Could you please look into tools that would keep them consistent?

Vim appears to have features for automatically filling paragraphs:  https://stackoverflow.com/questions/9509321/what-is-the-equivalent-of-emacss-fill-paragraph-in-macvim

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+                "The MovingWindow size is not a integer multiple of the distance."
+            )
+
+        if self._distance - period.total_seconds() / self._sampling_period != 0:
```
Maybe use `math.isclose()` here, or maybe better the alias for zero that we talked about it #337 ?

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        self._buffer = moving_window._buffer
+
+        # internally we are working with indexes of samples
+        self._sampling_period = self._moving_window.sampling_period.total_seconds()
```
Any reason to use a `float` here, for what I've seen, all operations you do are also supported by `timedelta` (even modulo), so you could just as well use `timedelta` internally, right?

matthias-wende-frequenz replied:
~All the internal operations are based on indexes and not on times any longer.~
You are right. I'll change it :)

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
````diff
@@ -0,0 +1,443 @@
+        ```python
+            feature_extractor = PeriodicFeatureExtractor(
+            moving_window = moving_window, # 28 days moving window
+            distance=timedelta(days=7),
````
``` suggestion
            moving_window=moving_window,  # 28 days moving window
            period=timedelta(days=7),

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+            ValueError: If the MovingWindow size is not a integer multiple of the period.
+        """
+        self._moving_window = moving_window
+        self._buffer = moving_window._buffer
```
I would probably also add a public `property` to access the underlying buffer of a `MovingWindow` if there is a case for it.

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        now = datetime.now(tzinfo=timezone.utc)
+
+        # create a daily weighted average for the next 24h
+        avg_24h = feature_extractor.avg(
+            now, now + timedelta(hours=24),
+            weights=[0.1, 0.2, 0.3, 0.4]
+        )
```
just some black-like formatting for the examples
``` suggestion
        avg_24h = feature_extractor.avg(
            now, 
            now+timedelta(hours=24),
            weights=[0.1, 0.2, 0.3, 0.4],
        )

leandro-lucarella-frequenz replied:
@mathias-baumann-frequenz maybe we should also run black, isort, pylint, etc. to all code examples in the future too :scream: (in a far future, not for now for sure)

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        self._sampling_period = self._moving_window.sampling_period.total_seconds()
+
+        # distance between two succeeding intervals in samples
+        self._distance = int(period.total_seconds() / self._sampling_period)
```
It is nice to document variables using docstrings, because then lsp can display it when we lookup/highlight instances of the variable:

``` suggestion
        self._distance = int(period.total_seconds() / self._sampling_period)
        """Distance between two succeeding intervals in samples."""

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        """
+        Plain python version of the average calculator.
+
+        This method avoids copying in any case but is 15 to 600
+        slower then the numpy version.
+
+        This method is only used in benchmarks and unit tests.
+
+        ONLY USE THIS IF YOU NEED TO STRICTLY AVOID COPIES!
```
You can use special blocks to highlight this, for example:

```suggestion
        Danger:
            This method avoids copying in any case but is 15 to 600
            slower then the numpy version.

            This method is only used in benchmarks and unit tests.

            **ONLY USE THIS IF YOU NEED TO STRICTLY AVOID COPIES!**
```

(you can use `Warning:` too)

Here is the output:

![image](https://user-images.githubusercontent.com/108664423/235879231-ff6d417e-ca01-4ab2-8369-f7e0cdd6e14e.png)

![image](https://user-images.githubusercontent.com/108664423/235879453-6467b921-14bb-436b-94a6-f9506e54553a.png)

The only downside is right now this is rendered collapsed, not expanded, but there is an [issue](https://github.com/mkdocstrings/mkdocstrings/issues/468) in `mkdocstrings` to fix this (and I think it should be a pretty easy so I guess it should come eventually).

leandro-lucarella-frequenz replied:
I just noticed this is private, so it won't be actually rendered, but in any case I think is good to always write docs as if they were public, in case sometime we decide to make it public, as it happened in this PR :D

sahas-subramanian-frequenz replied:
The `Note` and `Note further` in the class docstring can probably use this pattern as well?

leandro-lucarella-frequenz replied:
I missed those, but yeah, there is a `Note:` *admonition* too. You can see the full list here: https://squidfunk.github.io/mkdocs-material/reference/admonitions/#supported-types

leandro-lucarella-frequenz replied:
And now admonitions are expanded by default! :tada:

https://github.com/mkdocstrings/mkdocstrings/issues/468#issuecomment-1537489843

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+
+    def avg(
+        self, start: datetime, end: datetime, weights: List[float] | None = None
+    ) -> NDArray[Any]:
```
Shouldn't this return `NDArray[np.float_]`? I think it's not that terrible to do some type ignore in the internal methods but I would try to keep the public interface as strictly typed as possible.

leandro-lucarella-frequenz commented on benchmarks/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,213 @@
+    Returns:
+        The averaged window.
+    """
+    reshaped = feature_extractor._reshape_np_array(  # pylint: disable=W0212
```
Can you maybe replace the `pylint` numeric codes with the text representation (in this case `W0212` -> `protected-access`)? It is much harder to know what's being ignored otherwise.

leandro-lucarella-frequenz commented on benchmarks/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,213 @@
+    """
+    reshaped = feature_extractor._reshape_np_array(  # pylint: disable=W0212
+        window, window_size
+    )  # pylint: disable=W0212
```
Also do you really need this twice? :thinking:

matthias-wende-frequenz replied:
Just making sure pylint really listens! :D

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+                "The MovingWindow size is not a integer multiple of the distance."
+            )
+
+        if self._distance - period.total_seconds() / self._sampling_period != 0:
+            log.warning(
+                "The distance is not a multiple of the sampling period. "
+                "This might result in unexpected behaviour."
+            )
```
why not raise here too?  are there any use cases where this needs to be supported?

leandro-lucarella-frequenz replied:
+1, I actually thought the same when reading this but assumed there was. And if there is, it would be good to state it explicitly in a comment, so people looking at the code in the future don't ask themselves the same question.

leandro-lucarella-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -10,7 +10,18 @@
 ## New Features
 
-<!-- Here goes the main new features and examples or instructions on how to use them -->
+* The `MovingWindow` as well as the `OrderedRingBuffer` are having new public
+  methods that are returning the oldest and newest timestamp of all stored samples.
+
+* The `OrderedRingBuffer`s `wrap` and `normalize_timestamp` are now public.
+
+* The `PeriodicFeatureExtractor` has been added.
+
+  This is a tool to create certain profiles out of periodic reoccurring windows
+  inside a `MovingWindow`.
+
+  As an example one can create a daily profile of specific weekdays which will
+  be returned as numpy arrays.
```
Reformatted because of how notes are rendered in releases (GitHub keep the end of lines) and removed references to `OrderedRingBuffer` as it is not public (so users don't care about changes in it):

```suggestion
* The `MovingWindow` has new public methods that are returning the oldest and newest timestamp of all stored samples.

* The `PeriodicFeatureExtractor` has been added.

  This is a tool to create certain profiles out of periodic reoccurring windows inside a `MovingWindow`.

  As an example one can create a daily profile of specific weekdays which will be returned as numpy arrays.
```

leandro-lucarella-frequenz reviewed (commented):
OK, here is my review.  They are all small style-ish things, so LGTM in general.

I didn't get into the math to be honest because is not my strongest point, maybe someone from the @frequenz-floss/datasci-team can have a look at that. I also only looked at the benchmark and tests superficially, maybe someone else can review those in more detail, otherwise I can give it another go if you think it is important to review.

There is also a commit using an outdated class name: `Add tests for the PeriodicWindowProfiler`

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        """
+        Get the index of a timestamp relative to the oldest sample in the MovingWindow.
+
+        In other word consider an integer axis where the zero is defined
```
``` suggestion
        In other words, consider an integer axis where the zero is defined

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        # align timestamp to the sampling period
+        timestamp = self._buffer.normalize_timestamp(timestamp)
+
+        # distance between the input ts and the ts of oldest know samples (in samples)
```
``` suggestion
        # distance between the input ts and the ts of oldest known samples (in samples)

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        self, window: NDArray[np.float_] | MovingWindow, window_size: int
+    ) -> int:
+        """
+        Get the number of windows that are fully contained in the MovingWindow.
```
looking at the code,  it looks like this returns the number of windows that "can" be contained in the MovingWindow, no?  There are no windows when no data has been inserted, for example.

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        Returns:
+            The index of the timestamp in the MovingWindow.
+        """
+        return self._buffer.datetime_to_index(timestamp)
```
this function is just getting forwarded, to `_buffer.datetime_to_index`, and is just getting called once.  Maybe it can be dropped.

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        """
+        num_windows = math.floor(len(window) / self._distance)
+        if len(window) - num_windows * self._distance >= window_size:
+            num_windows += 1
```
maybe we need a comment that says `len(window)` returns the number of samples, not the size of the buffer.  Also an explanation for what this means: `len(window) - num_windows * self._distance >= window_size`

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        weights: List[float] | None = None,
+    ) -> NDArray[np.float_]:
+        """
+        Plain python version of the average calculator.
```
to make it sound more imperative.

``` suggestion
        """
        Calculate average using a plain Python implementation.

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+
+        # Case 1:
+        if len(array) - num_windows * self._distance == 0:
+            log.debug("Case 1:")
```
Not sure these log lines are helpful in a large app, without any additional info.

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        Plain python version of the average calculator.
+
+        This method avoids copying in any case but is 15 to 600
+        slower then the numpy version.
```
``` suggestion
        This method avoids copying in all cases, but is 15 to 600 times
        slower than the numpy version.

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+
+        for i in range(window_size):
+            assert num_windows * (self._distance) - len(window) <= self._distance
+            summe = 0
```
:de:

matthias-wende-frequenz replied:
Visual Basic style!

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+                else:
+                    summe += weights[j] * window[i + (j * (self._distance))]
+
+            if weights is None:
+                res[i] = summe / num_windows
+            else:
+                res[i] = summe / np.sum(weights)
```
You also need to check `weights` is an empty list, before sending to `np.sum`,  you can change the check such that it would pass if weights is None, an empty list or set etc.

``` suggestion
            if not weights:
                res[i] = summe / num_windows
            else:
                res[i] = summe / np.sum(weights)

sahas-subramanian-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        # Note that the returned value is a index not a timestamp
+        rel_start_sample = self._timestamp_to_rel_index(start) % self._distance
+        rel_end_sample = self._timestamp_to_rel_index(end) % self._distance
+        # fix the position if modulo period the end_sample
+        # is smaler then the start sample position
```
what does "if modulo period the end_sample" mean?

``` suggestion
        # fix the position if modulo period the end_sample
        # is smaller then the start sample position

matthias-wende-frequenz replied:
it means the `rel_next_position % self.period`

sahas-subramanian-frequenz commented on benchmarks/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,213 @@
+    return np.average(reshaped[:, :window_size], axis=0)  # type: ignore[no-any-return]
+
+
+def _calculate_avg_window_py(
+    feature_extractor: PeriodicFeatureExtractor,
+    window: NDArray[np.float_],
+    window_size: int,
+    weights: List[float] | None = None,
+) -> NDArray[np.float_]:
+    """
+    Plain python version of the average calculator.
+
+    This method avoids copying in any case but is 15 to 600
+    slower then the numpy version.
+
+    This method is only used in these benchmarks.
```
There's an implementation of this method in the `PeriodicFeatureExtractor` class,  with the same code.  Can we drop this function in the benchmark file and use the one from the class?  If not,  maybe useful to document why not.

matthias-wende-frequenz replied:
Ups. That should not be in the `PeriodicFeatureExtractor` any longer. I'll will remove from there.

sahas-subramanian-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,137 @@
+    """
+    Initialize a PeriodicFeatureExtractor with a `MovingWindow` that contains the data.
+
+    Parameters:
+        data: The data that is pushed into the moving window.
+        period: The distance between two successive windows.
```
``` suggestion
    Args:
        data: The data that is pushed into the moving window.
        period: The distance between two successive windows.

sahas-subramanian-frequenz reviewed (commented):
Many commit messages still say "TimeseriesAvg" and "PeriodicWindowProfiler" instead of "PeriodicFeatureExtractor"

cwasicki commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,137 @@
+    data: List[float] = [1, 2, 2.5, 1, 1, 1, 2, 2, 1, 1, 1, 1, 1, 1, 1]  # type: ignore[no-redef]
+
+    feature_extractor = await init_feature_extractor(data, timedelta(seconds=5))
+    assert np.allclose(feature_extractor.avg(start, end), [1.5, 1.5])
```
Added more tests in #375

cwasicki commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,443 @@
+        Args:
+            start: The start of the window to average over.
+            end: The end of the window to average over.
+            weights: The weights to use for the average calculation.
```
From this it's not clear in which order the weights are applied (i.e. oldest first)

matthias-wende-frequenz replied:
yes it's oldest first as dictated by numpy :). Will add a comment

leandro-lucarella-frequenz commented on src/conftest.py:
```diff
@@ -122,6 +122,8 @@ def evaluate(self, example: Example) -> None | str:
         import_header.append(
             path_to_import_statement(Path(os.path.relpath(example.path)))
         )
+        # Add microgrid as default import
+        import_header.append("from frequenz.sdk import microgrid")
```
I'm not sure if it is a good idea to add this, is just one extra line for examples and it makes examples more complete.

leandro-lucarella-frequenz reviewed (commented):
Just one more small comment, I started to review again but then noticed a new review wasn't requested yet so not sure if you are still working on the PR... Please request a review again when it is ready for a new round.

cwasicki commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,392 @@
+        self._sampling_period = self._moving_window.sampling_period.total_seconds()
+        """The sampling_period as integer to use it for indexing of samples."""
+
+        self._period = int(period.total_seconds() / self._sampling_period)
```
Does that work correctly if period is not a multiple of sampling period?

cwasicki replied:
oh, I see you check this below. For unexpected behavior I would raise an exception or do you see valid use-cases for this?

matthias-wende-frequenz replied:
I didn't want to add unnecessary restrictions and I haven't thought it fully through. I suggest we address this by adding tests in a follow up PR.

cwasicki commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,392 @@
+        Returns:
+            The number of windows that are fully contained in the MovingWindow.
+        """
+        num_windows = math.floor(len(window) / self._period)
```
Could be `//` instead of floor (given it's positive)

cwasicki commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,392 @@
+            The number of windows that are fully contained in the MovingWindow.
+        """
+        num_windows = math.floor(len(window) / self._period)
+        if len(window) - num_windows * self._period >= window_size:
```
When can this happen?

matthias-wende-frequenz replied:
This can't happen any longer with the current restrictions. This function was actually meant to work in a more generic situation. I'll comment that code out and leave it was a comment for future reference.

matthias-wende-frequenz replied:
I moved that function to the benchmarks as at hasn't been used in the code at all.

cwasicki commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,392 @@
+        1. the array size is a multiple of window_size + period,
+           i.e. num_windows is integer and we can simply reshape
+        2. the array size is NOT a multiple of window_size + period and
+           it has length n * (window_size + period) + m, where m < window_size
```
`window_size+period` -> `period`?

cwasicki commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,392 @@
+        2. the array size is NOT a multiple of window_size + period and
+           it has length n * (window_size + period) + m, where m < window_size
+        3. the array size is NOT a multiple of window_size + period and
+           it has length n * (window_size + period) + m, where m >= window_size
```
Can this happen?

matthias-wende-frequenz replied:
only case 1 can happen in the current implementation. But afair all three cases are covered in the benchmarks or the tests. I'll leave it for future work, when we remove the condition and add a comment.

matthias-wende-frequenz commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,392 @@
+
+        # distance between the input ts and the ts of oldest know samples (in samples)
+        dist_to_oldest = int(
+            abs((timestamp - self._moving_window.time_bound_oldest).total_seconds())
```
Using abs is a bug. Using modulo arithmetic alone is enough and thus it's fine to return a negative value here.

matthias-wende-frequenz committed:
Update release notes

Signed-off-by: Matthias Wende <matthias.wende@frequenz.com>

shsms commented on src/frequenz/sdk/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,437 @@
+"""
+A class that extracts features out of historical data.
+
+It creats a profile from periodically occurring windows in a buffer of
```
```suggestion
It creates a profile from periodically occurring windows in a buffer of

shsms reviewed (dismissed):
🎉

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+import numpy as np
+from numpy.typing import NDArray
+
+from frequenz.sdk._internal._math import is_close_to_zero
```
```suggestion
from .._internal._math import is_close_to_zero
```

leandro-lucarella-frequenz commented on RELEASE_NOTES.md:
```diff
@@ -28,7 +28,15 @@ This release drops support for Python versions older than 3.11.
-<!-- Here goes the main new features and examples or instructions on how to use them -->
+* The `MovingWindow` as well as the OrderedRingBuffer are having new public methods that are returning the oldest and newest timestamp of all stored samples.
+
+* The `OrderedRingBuffer`s `wrap` and `normalize_timestamp` are now public.
```
This is not being exposed, so technically it shouldn't be in the release notes, as users are not supposed to use it directly.

matthias-wende-frequenz replied:
True

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+            _logger.warning(
+                "The period is not a multiple of the sampling period. "
+                "This might result in unexpected behaviour."
+            )
```
There was a pending question about if we want to allow this at all or not (i.e. raising an exception or leaving it only as a warning).

matthias-wende-frequenz replied:
To be honest I haven't fully thought these cases through but I think it'll work. Since I didn't want to add any unnecessary restrictions I left the warning and decided to add follow up unit tests to get some more insights.

cwasicki replied:
I think if we don't know how it behaves we should raise, we can still change it when we have more insights.

matthias-wende-frequenz replied:
Fair enough! Will change to raise :).

leandro-lucarella-frequenz commented on tests/timeseries/test_moving_window.py:
```diff
@@ -133,3 +143,6 @@ async def test_resampling_window(fake_time: time_machine.Coordinates) -> None:
     assert len(window) == window_size / output_sampling
     for value in window:  # type: ignore
         assert 4.9 < value < 5.1
+
+
+asyncio.run(test_access_window_by_int_slice())
```
Why this? `pytest` should properly run `async` tests in a async context automatically.

```suggestion
```

matthias-wende-frequenz replied:
removed. good catch!

leandro-lucarella-frequenz commented on tests/timeseries/test_moving_window.py:
```diff
@@ -14,6 +14,7 @@
 from frequenz.sdk.timeseries import Sample
+from frequenz.sdk.timeseries._base_types import UNIX_EPOCH
 from frequenz.sdk.timeseries._moving_window import MovingWindow
 from frequenz.sdk.timeseries._resampling import ResamplerConfig
```
Test should try to use public imports when possible. I think all symbols in here are available simply in `frequenz.timeseries`.
```suggestion
from frequenz.sdk.timeseries import Sample, UNIX_EPOCH, MovingWindow, ResamplerConfig
```

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""
+A class that extracts features out of historical data.
```
I would remove the "a class that", maybe something like:

```suggestion
An extractor to get features out of historical data.
```

Also, please bear in mind that this module is **private**, so all this documentation won't be shown in the rendered documentation in the website. I'd suggest moving all the important documentation to the class instead.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+@dataclass(frozen=True)
+class RelativePositions:
+    """
+    A dataclass that holds the relative positions of a window in a buffer.
```
Same here, maybe:

```suggestion
    The relative positions of a window in a buffer.
```

BTW, positions of what? By reading the class doc for me it is not clear what this is actually for.

matthias-wende-frequenz replied:
I've reworked the documentation by adding more context.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+    """
+
+    start: int
+    end: int
+    next: int
```
Ideally these should be documented too.

```suggestion
    start: int
    """Whatever."""

    end: int
    """Whatever."""

    next: int
    """Whatever."""
```

matthias-wende-frequenz replied:
Done but I changed Whatever to something meaningful :p.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+            assert isinstance(window_end, np.ndarray)
+            window_array = np.concatenate((window_start, window_end))
+        else:
+            # ignoring the type check because we know that the returned type is an ndarray
```
Is this a left-over of a previous version? There is not type ignore here...

matthias-wende-frequenz replied:
probably

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+            The averaged timeseries window.
+        """
+        (reshaped, window_size) = self._get_reshaped_np_array(start, end)
+        return np.average(  # type: ignore[no-any-return]
```
Maybe it was for this line?

matthias-wende-frequenz replied:
probably not :D

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,414 @@
+
+    @property
+    def _buffer(self) -> OrderedRingBuffer[NDArray[np.float64]]:
+        return self._moving_window._buffer  # pylint: disable=protected-access
```
Any particular reason why not just make the `MovingWindow` buffer a public property too?

matthias-wende-frequenz replied:
You mean in the `MovingWindow`? I'd rather prefer to keep it hidden for users.

leandro-lucarella-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
Not for this PR, but as a general thing. I think we should start mocking more in **unit tests**.  These tests are not testing the feature extractor in isolation, they are also testing the moving window, and the moving window is also testing the ring buffer and on top of that, it is using a channel, and senders and receivers, so there is quite a bit going on here.

This tend to make tests much slower and sometimes more complicated, or duplicate a lot of testing from the used classes.  Also, it happened to me that working on some component suddenly test for a completely different component start to fail, which is confusing and distracting.

I think having **integration tests** is also important, but I feel we are mixing them too much, so in the future I'd try to make a more clear boundary between unit tests and integration tests, specially as we build higher and higher abstractions the problem intensifies if we keep testing the whole stack when testing the higher level abstractions in unit tests.

matthias-wende-frequenz replied:
Agreed!

leandro-lucarella-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,140 @@
+        PeriodicFeatureExtractor
+    """
+    window, sender = init_moving_window(timedelta(seconds=len(data)))
+    await push_lm_data(sender, data)
```
Just for the records, I had to go look into the moving window test to see what "lm" stands for, it would be more clear to call it `push_logical_meter_data()` :)

leandro-lucarella-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,140 @@
+from frequenz.channels import Broadcast
+
+from frequenz.sdk.timeseries import MovingWindow, PeriodicFeatureExtractor, Sample
+from frequenz.sdk.timeseries._base_types import UNIX_EPOCH
```
```suggestion
from frequenz.sdk.timeseries import MovingWindow, PeriodicFeatureExtractor, Sample, UNIX_EPOCH
```

leandro-lucarella-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,140 @@
+    lm_chan = Broadcast[Sample]("lm_net_power")
+    # monkey patch the MovingWindow class to have a fixed size such that
+    # it can be initialized with a distance bigger then the window size.
+    MovingWindow.__len__ = lambda self: period  # type: ignore[method-assign]
```
You should use `with mock.patch(...)` or the `mocker` *fixture* so the `MovingWindow.__len__` is restored after the test, otherwise you can run to really weird shit.

matthias-wende-frequenz replied:
Actually in that case the patching is not needed any longer, since the semantics of period have changed.

leandro-lucarella-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,140 @@
+    data = np.array([2, 2.5, 1, 1, 1, 2])
+    feature_extractor = await init_feature_extractor_no_data(4)
+    window_size = 2
+    reshaped = feature_extractor._reshape_np_array(  # pylint: disable=W0212
```
Please use the symbolic `pylint` check names (`protected-access` in this case)

matthias-wende-frequenz replied:
I thought I changed all these comments.

leandro-lucarella-frequenz commented on tests/timeseries/test_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,140 @@
+    window_size = 2
+    reshaped = feature_extractor._reshape_np_array(  # pylint: disable=W0212
+        data, window_size
+    )  # pylint: disable=W0212
```
Also do they really need to be specified twice?

matthias-wende-frequenz replied:
no they don't. Fixed!

leandro-lucarella-frequenz reviewed (commented):
Quite a few comments but most of them super minor. The only 2 comments that are important to me are:

1. Patching of `MovingWindow.__len__` (and not restoring it)
2. The `asyncio.run()` in the test

And there is also the unit tests doing to little mocking, which is quite important IMHO, but again, I don't expect that to be changed in this PR as it is a much bigger topic and we shouldn't block this PR until we address it.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/_periodic_feature_extractor.py:
```diff
@@ -0,0 +1,419 @@
+    end: int
+    """The relative end position of the window."""
+    next: int
+    """The relative position of the next incoming sample sample."""
```
```suggestion
    """The relative position of the next incoming sample."""
```

matthias-wende-frequenz replied:
That's intended! Having it twice makes a bigger impact :D.

leandro-lucarella-frequenz commented on benchmarks/timeseries/periodic_feature_extractor.py:
```diff
@@ -0,0 +1,235 @@
+    lm_chan = Broadcast[Sample]("lm_net_power")
+    # monkey patch the MovingWindow class to have a fixed size such that
+    # it can be initialized with a period bigger then the window size.
+    MovingWindow.__len__ = lambda self: period  # type: ignore[method-assign]
```
The patching is still there!

matthias-wende-frequenz replied:
weird I can't find it

matthias-wende-frequenz replied:
ah damn found it. That was done in 2 places.

leandro-lucarella-frequenz reviewed (commented):
Maybe you forgot to push, but the patching of the `MovingWindow.__len__` is still there. And a new typo. Other than that, LGTM!
